### PR TITLE
renamed GalleryHasMedia to GalleryItem

### LINKED ISSUE #828: [DX] would like to rename the GalleryHasMedia object
to ```GalleryMediaItem``` ... an GalleryItem which holds the reference to the Media.

What do you think?

another reason for me is, that this name is **countable** -> GalleryMediaItems -> many items, where every item contains one Media

vs.

GalleryHasMedias..... 

If you like it, i will provide a PR.

I know this is a BC break.

best Oskar

### PR DESCRIPTION
**Pull request by OskarStark**
Closes #828 

### Changelog

```markdown
### Changed
- Renamed `GalleryHasMedia` to `GalleryItem` and all controllers, interfaces and actions accordingly
- Renamed `setGalleryHasMedias` to `setGalleryItems` in `Gallery`
- Renamed `getGalleryHasMedias` to `getGalleryItems` in `Gallery`
- Renamed `addGalleryHasMedias` to `addGalleryItem` in `Gallery`
- Renamed `$galleryHasMedias` to `$galleryItems` in `Gallery`
- Renamed `GalleryHasMediaAdmin` to `GalleryItemAdmin`
- Renamed `$galleryHasMediaClass` to `$galleryItemClass` in `GalleryController`
- Renamed `getGalleryGaleryhasmediasAction` to `getGalleryGalleryItemAction` in `GalleryController`
- Renamed `postGalleryMediaGalleryhasmediaAction` to `postGalleryMediaGalleryItemAction` in `GalleryController`
- Renamed `handleWriteGalleryhasmedia` to `handleWriteGalleryItem` in `GalleryController`
- Renamed `putGalleryMediaGalleryhasmediaAction` to `putGalleryMediaGalleryItemAction` in `GalleryController`
- Renamed `deleteGalleryMediaGalleryhasmediaAction` to `deleteGalleryMediaGalleryItemAction` in `GalleryController`
- Changed route name from `sonata_media_api_form_gallery_has_media ` to `sonata_media_api_form_gallery_item`
- Renamed `GalleryHasMediaInterface` to `GalleryItemInterface`
- Renamed config key from `gallery_has_media` to `gallery_item` under `sonata_media.class`
- Renamed `BaseGalleryHasMedia` to `BaseGalleryItem`
- Renamed `ApiGalleryHasMediaType` to `ApiGalleryItemType`
- Changed `GalleryInterface` and renamed `setGalleryHasMedias`, `getGalleryHasMedias` and `addGalleryHasMedias` to `setGalleryItems`, `getGalleryItems` and `addGalleryItem`
- Renamed `reorderGalleryHasMedia` to `reorderGalleryItems` in `BaseGallery`
- Renamed `BaseGalleryHasMediaRepository` to `BaseGalleryItemRepository`
- Renamed `GalleryHasMediaRepository` to `GalleryItemRepository`
- Renamed `form.type` alias from `sonata_media_api_form_gallery_has_media` to `sonata_media_api_form_gallery_item`
- Renamed PHP files, skeletons and schema XML's accordingly to their class/entity names
- Renamed `serializer/Model.GalleryHasMedia.xml` to `serializer/Model.GalleryItem.xml`
- Renamed `translation keys from `form.label_gallery_has_medias` to `form.label.gallery_items`
- Changed `Resources/views/Gallery/view.html.twig`
```

### Subject

This change is to huge to keep it BC, so i will schedule this for the next major release.

@sonata-project/contributors please review this change, thank you!

### CODE REVIEW

**Comment by OskarStark** [comment]
Can we merge this in 3.1 ? @Soullivaneuh

**Comment by OskarStark** [comment]
ping @greg0ire can you please review?

File: Block/GalleryBlockService.php [hunk 1]
```diff
@@ -221,22 +221,22 @@ public function preUpdate(BlockInterface $block)
...
 
             $elements[] = array(
-                'title'     => $galleryHasMedia->getMedia()->getName(),
-                'caption'   => $galleryHasMedia->getMedia()->getDescription(),
-                'type'      => $type,
-                'media'     => $galleryHasMedia->getMedia(),
+                'title'   => $galleryItem->getMedia()->getName(),
+                'caption' => $galleryItem->getMedia()->getDescription(),
+                'type'    => $type,
+                'media'   => $galleryItem->getMedia(),
```
Comment (greg0ire) [inline Block/GalleryBlockService.php]: Didn't we stop aligning things?

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,10 @@ A [BC BREAK] means the update will break the project for many reasons:
 * new dependencies
 * class refactoring
 
+### 2016-02-14
+
+* [BC BREAK] Renamed `GalleryHasMedia` to `GalleryItem`. All Actions, Controllers, Interfaces and anything related to this is renamed accordingly.
```
Comment (greg0ire) [inline CHANGELOG.md]: This line is too long.

File: Controller/Api/GalleryController.php [hunk 3]
```diff
@@ -53,22 +53,22 @@ class GalleryController
...
      * Constructor.
      *
      * @param GalleryManagerInterface $galleryManager
      * @param MediaManagerInterface   $mediaManager
      * @param FormFactoryInterface    $formFactory
-     * @param string                  $galleryHasMediaClass
+     * @param string                  $galleryItemClass
      */
-    public function __construct(GalleryManagerInterface $galleryManager, MediaManagerInterface $mediaManager, FormFactoryInterface $formFactory, $galleryHasMediaClass)
+    public function __construct(GalleryManagerInterface $galleryManager, MediaManagerInterface $mediaManager, FormFactoryInterface $formFactory, $galleryItemClass)
```
Comment (greg0ire) [inline Controller/Api/GalleryController.php]: Please linebreak this

File: Controller/Api/GalleryController.php [hunk 4]
```diff
@@ -163,24 +163,24 @@ public function getGalleryAction($id)
...
     public function getGalleryMediasAction($id)
     {
-        $ghms = $this->getGallery($id)->getGalleryHasMedias();
+        $galleryItems = $this->getGallery($id)->getGalleryItems();
 
         $media = array();
-        foreach ($ghms as $ghm) {
-            $media[] = $ghm->getMedia();
+        foreach ($galleryItems as $galleryItem) {
+            $media[] = $galleryItem->getMedia();
```
Comment (greg0ire) [inline Controller/Api/GalleryController.php]: Lol that's a bit more meaningful :)

File: Controller/Api/GalleryController.php [hunk 5]
```diff
@@ -318,41 +318,41 @@ public function putGalleryMediaGalleryhasmediaAction($galleryId, $mediaId, Reque
...
-     * @param Request                  $request
+     * @param GalleryInterface     $gallery
+     * @param MediaInterface       $media
+     * @param GalleryItemInterface $galleryItem
+     * @param Request              $request
      *
      * @return FormInterface
      */
-    protected function handleWriteGalleryhasmedia(GalleryInterface $gallery, MediaInterface $media, GalleryHasMediaInterface $galleryHasMedia = null, Request $request)
+    protected function handleWriteGalleryhasmedia(GalleryInterface $gallery, MediaInterface $media, GalleryItemInterface $galleryItem = null, Request $request)
```
Comment (greg0ire) [inline Controller/Api/GalleryController.php]: https://www.youtube.com/watch?v=Z6_ZNW1DACE

**Comment by mvhirsch** [comment]
:+1: for such a change, this will finally clarify what my ```$gallery->getGalleryHasMedias()``` (now ```$gallery->getGalleryItems()```) is supposed to do

File: UPGRADE-3.0.md [hunk 6]
```diff
@@ -0,0 +1,10 @@
+UPGRADE FROM 2.x to 3.0
+=======================
+
+### Dependencies
+
+You will need to follow the dependencies upgrade instructions.
+
+## GalleryHasMedia
+
+The `GalleryHasMedia` has been renamed to `GalleryItem`, please adjust your code and be sure to move your assets to the correct database table.
```
Comment (greg0ire) [inline UPGRADE-3.0.md]: `gqq` in vim

**Comment by greg0ire** [comment]
Huge refactoring indeed! Congrats!

File: Block/GalleryBlockService.php (hunk 1 above)
Comment (OskarStark) [inline Block/GalleryBlockService.php]: i rebased, but this is old and should then be fixxed by Style CI IMO

Comment (greg0ire) [inline Block/GalleryBlockService.php]: Ok fine :)

**Comment by Soullivaneuh** [comment]
> Can we merge this in 3.1 ? @Soullivaneuh

Yes after I push 3.0. But hits have to be BC.

**Comment by OskarStark** [comment]
> Yes after I push 3.0. But it has to be BC.

then lets merge it in 4.0, this is a massive change, all classes needs to be handeld...

File: UPGRADE-3.0.md (hunk 6 above)
Comment (OskarStark) [inline UPGRADE-3.0.md]: ?? what do you mean? :)

Comment (greg0ire) [inline UPGRADE-3.0.md]: That it's too long. If you open the file with vim, and type `10Ggqq`, it will be fixed :)

File: Controller/Api/GalleryController.php (hunk 4 above)
Comment (OskarStark) [inline Controller/Api/GalleryController.php]: yes 😄

**Comment by Soullivaneuh** [comment]
You rename the class `GalleryHasMediaAdmin`, why not simply keep it and mark it as deprecated?

**Comment by OskarStark** [comment]
Really? IMO we need to deprecate all (Api-)Controller classes too and maybe something else.

and i don't know what will happen if you upgrade and upgrade your schema, the new table name shouldn't match the new generated one, shouldn't it?

**Comment by Soullivaneuh** [comment]
According to the new [Sonata version management](https://github.com/sonata-project/SonataAdminBundle/issues/3053) and next major [release plan](https://github.com/sonata-project/SonataAdminBundle/issues/3731), this project has been refactored regarding branching and versioning.

If you see this message, your PR concerns a patch or a minor release and is not targeting the right branch.

So I'm closing this one, but don't see it as a refusal. If you think your work is still relevant and want to continue, feel free to reopen it on the right branch (e.g. the default one).

Regards.

**Comment by OskarStark** [comment]
This is for the next major

**Comment by Soullivaneuh** [comment]
According to the new [Sonata version management](https://github.com/sonata-project/SonataAdminBundle/issues/3053) and next major [release plan](https://github.com/sonata-project/SonataAdminBundle/issues/3731), this project has been refactored regarding branching and versioning.

If you see this message, your PR concerns a patch or a minor release and is not targeting the right branch.

So I'm closing this one, but don't see it as a refusal. If you think your work is still relevant and want to continue, feel free to reopen it on the right branch (e.g. the default one).

Regards.

**Comment by Soullivaneuh** [comment]
@OskarStark Can this be done on 3.x with a BC way?

**Comment by OskarStark** [comment]
> @OskarStark Can this be done on 3.x with a BC way?

As i explained here: https://github.com/sonata-project/SonataMediaBundle/pull/948#issuecomment-219012946

No

File: Block/GalleryBlockService.php (hunk 1 above)
Comment (OskarStark) [inline Block/GalleryBlockService.php]: done

**Comment by OskarStark** [comment]
rebased

File: Controller/Api/GalleryController.php (hunk 3 above)
Comment (OskarStark) [inline Controller/Api/GalleryController.php]: we won't

Comment (greg0ire) [inline Controller/Api/GalleryController.php]: I know T___T

**Comment by greg0ire** [comment]
> Could you please rebase your PR and fix merge conflicts?

SonataCI ? Is that you?

**Comment by Soullivaneuh** [comment]
Could you please rebase your PR and fix merge conflicts?

**Comment by Soullivaneuh** [comment]
Oo

**Comment by OskarStark** [comment]
i will rebase again in evening

**Comment by greg0ire** [comment]
> i will rebase again in evening
> Could you please rebase your PR and fix merge conflicts?

That's nice, but could you please rebase ?

**Comment by OskarStark** [comment]
done @greg0ire

**Comment by OskarStark** [comment]
StyleCi could not be executed 😢 

ping @Soullivaneuh

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,6 @@
 # Change Log
 All notable changes to this project will be documented in this file.
 This project adheres to [Semantic Versioning](http://semver.org/).
+
+CHANGELOG
+=========
```
Comment (core23) [inline CHANGELOG.md]: Why did you add this?

Comment (OskarStark) [inline CHANGELOG.md]: good catch, maybe i merged sth wrong!

thank you 👍

**Comment by OskarStark** [comment]
Shall i diretly extend `AbstractAdmin` or do we want to do this in a separate PR?

**Comment by core23** [comment]
> Shall i directly extend AbstractAdmin or do we want to do this in a separate PR?

We should do this in a separate PR. This topic is about renaming `GalleryHasMedia` to `GalleryItem`.

**Comment by OskarStark** [comment]
> We should do this in a separate PR. This topic is about renaming GalleryHasMedia to GalleryItem.

Thanks for your feedback, thats also my opinion 👍

**Comment by OskarStark** [comment]
StyleCI green ✅

**Comment by OskarStark** [comment]
puh... i updated the Changelog, maybe we can use this to put in in the UPGRADE.md file

**Comment by OskarStark** [comment]
travis is green, but SonataCI Quick Check isn't!

/cc @greg0ire @Soullivaneuh

**Comment by greg0ire** [comment]
> There were 161 warnings:
>
> 1) Sonata\MediaBundle\Tests\Block\Breadcrumb\BreadcrumbTest::testBlockService
PHPUnit_Framework_TestCase::getMock() is deprecated, use PHPUnit_Framework_TestCase::createMock() or PHPUnit_Framework_TestCase::getMockBuilder() instead

WAT

**Comment by mvhirsch** [comment]
@greg0ire PHPUnit 5.4 added the new ```createMock()``` method and deprecated ```getMock()```, see here: https://github.com/sebastianbergmann/phpunit/wiki/Release-Announcement-for-PHPUnit-5.4.0

**Comment by OskarStark** [comment]
> @greg0ire PHPUnit 5.4 added the new createMock() method and deprecated getMock(), see here: https://github.com/sebastianbergmann/phpunit/wiki/Release-Announcement-for-PHPUnit-5.4.0

yes but that didn't explain why travis is fine and Quick Check not

We should rename this in a separate PR

**Comment by greg0ire** [comment]
> the __construct() and __clone() methods of the original class are not executed

@mvhirsch well that's a huge improvement!

> yes but that didn't explain why travis is fine and Quick Check not

Is the phpunit version the same?

**Comment by OskarStark** [comment]
>> the __construct() and __clone() methods of the original class are not executed
>
> @mvhirsch well that's a huge improvement!

i cannot find this comment in this PR, can you help me?

> Is the phpunit version the same?

I don't know

**Comment by greg0ire** [comment]
> i cannot find this comment in this PR, can you help me?

It's not in the PR, it's in the link @mvhirsch provided

**Comment by OskarStark** [comment]
> It's not in the PR, it's in the link @mvhirsch provided

got it, thx!

**Comment by greg0ire** [comment]
From http://91.121.3.21/job/sonata-ci-bundle-media/406/console :

> PHPUnit 5.4.2 by Sebastian Bergmann and contributors.

**Comment by greg0ire** [comment]
Oh wait Travis has the same problems in fact : https://travis-ci.org/sonata-project/SonataMediaBundle/jobs/135807506#L623

**Comment by greg0ire** [comment]
So it must simply be that SonataCI is more strict `--fail-on-warning` config flag ? @rande @Soullivaneuh

**Comment by OskarStark** [comment]
rebased

**Comment by OskarStark** [comment]
rebased again!

Can we please merge this?

**Comment by greg0ire** [comment]
Why is it marked as "in progress"?

**Comment by OskarStark** [comment]
oops, sorry

File: UPGRADE-4.0.md [hunk 8]
```diff
@@ -18,3 +18,7 @@ If you have implemented a custom model, you must adapt the signature of the foll
  * `GalleryHasMediaInterface::getId`
  * `GalleryInterface::getId`
  
+## Renamed GalleryHasMedia to GalleryItem
+
+All Actions, Controllers, Interfaces and anything related to this is renamed accordingly.
+>>>>>>> renamed GalleryHasMedia to GalleryItem
```
Comment (greg0ire) [inline UPGRADE-4.0.md]: Hum…

Comment (OskarStark) [inline UPGRADE-4.0.md]: oO

Comment (OskarStark) [inline UPGRADE-4.0.md]: fixed

**Comment by greg0ire** [comment]
Thanks for the huge work @OskarStark !

**Comment by OskarStark** [comment]
> Thanks for the huge work @OskarStark !

Thank you for your feedback @greg0ire 👍

**Comment by core23** [comment]
@OskarStark Is there a reason why we kept `Modeel.GalleryHasMedia.xml`?

**Comment by OskarStark** [comment]
not really, did i forget it?

**Comment by core23** [comment]
The file still exists, I would say yes :trollface:

**Comment by OskarStark** [comment]
i created a PR here: #1094
